torch/distributed/_composable/fsdp/_fsdp_collectives.py

```diff
@@ -313,6 +350,14 @@ def foreach_reduce(
         reduce_scatter_group, all_reduce_group, reduce_dtype
     )
     world_size = reduce_scatter_group.size()
+    for i, (fsdp_param, unsharded_grad) in enumerate(zip(fsdp_params, unsharded_grads)):
+        if (shard_dim := fsdp_param.fsdp_placement.dim) == 0:
+            continue
+        assert (
+            unsharded_grad.size(shard_dim) % world_size == 0
+        ), f"Shard({shard_dim}) requires even sharding: {unsharded_grad.size()=} {world_size=}"
+        chunks = torch.chunk(unsharded_grad, world_size, dim=shard_dim)
+        unsharded_grads[i] = torch.cat(chunks, dim=0)
```

Comment:
are we chunk + cat to make unsharded_grads contiguous? does the `.cat` trigger copies on gpu?

Reply:
We have to chunk -> cat to do a data shuffle. If I have some time, I may try to draw some diagrams to add to the PR description.